src/utility-types.d.ts

```diff
@@ -0,0 +1,11 @@
+/**
+ * These extra type definitions expand the utility types provided by TypeScript
+ */
+
+type RecursivePartial< T > = {
+	[ P in keyof T ]?: T[ P ] extends ( infer U )[]
+		? RecursivePartial< U >[]
+		: T[ P ] extends object | undefined
+		? RecursivePartial< T[ P ] >
+		: T[ P ];
+};
```

Comment:
@fluiddot alright, I'm going to "bite." 😄 What is this doing and why is it necessary? 😆 

Joking aside, I'd love to understand this logic. It's a bit cryptic to me. Would you be willing to share any insight you have? Thanks! 🙇🏻

Reply:
@dcalhoun I'm glad you asked about this 😃 .

This type is basically an extension of [Partial's TypeScript utility type](https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype) to make all nested properties optional. `Partial` only sets as optional the root ones. As an example, here's the `UserData` type:
```
export interface UserData {
	sites: SiteDetails[];
	snapshots: Snapshot[];
	devToolsOpen?: boolean;
	authToken?: StoredToken;
	onboardingCompleted?: boolean;
	lastBumpStats?: {
		[ group: string ]: {
			[ stat: string ]: number;
		};
	};
	promptWindowsSpeedUpResult?: PromptWindowsSpeedUpResult;
}
```

If you make `Partial<UserData>`, all the above properties will be optional:
```
{
    sites?: SiteDetails[] | undefined;
    snapshots?: Snapshot[] | undefined;
    devToolsOpen?: boolean | undefined;
    authToken?: StoredToken | undefined;
    onboardingCompleted?: boolean | undefined;
    lastBumpStats?: {
        ...;
    } | undefined;
    promptWindowsSpeedUpResult?: PromptWindowsSpeedUpResult | undefined;
}
```

However, for instance, the nested properties of `SiteDetails` like [`SiteDetails.id`](https://github.com/Automattic/studio/blob/f68e2f0ac693eaacb75964699fa12cd34e316110/src/ipc-types.d.ts#L17) won't be optional. I found this situation while writing [this unit test](https://github.com/Automattic/studio/pull/262/commits/bbb5b334564c67abba12436aaf826342a0194e61), and needed to mock the persisted user data. In my case, I only wanted to define part of the site details. That's why I introduced this extra utility type. I presume that we might need it in the future in similar cases, mostly when adding unit tests.